### DISCUSSION TITLE
Slow Sending Rate 🐌

### DISCUSSION THREAD

**Discussion by ffsandi** [postalserver/postal#1461] [Help with using Postal]
Dear teams,

I would like to ask why I only get about 10-35 emails/ minute when sending using SMTP from my Mautic.

I currently run on VPS with 4 core 8GB ram, and I also already try with 2 different VPS providers with the same specs.

I try several combinations, from 10 to 40 workers, and it doesn't give the significant result that I expect. 

40 workers consuming around 50% ram, but not increasing the sending rate speed.

I see another member here could hit 900 emails/minute.

Do you have any troubleshooting guide to solve this problem?

Because it will need more than 2 days to finish sending emails to over 50K subscribers, it feels that I could get older faster waiting for this to finish 🤣

Thanks.

@adamcooke @willpower232

**Comment by willpower232** [comment]
That is quite slow, is it possible that Mautic is having trouble sending that many messages quickly? Is your database on the same server as Postal? It might be fighting for memory. Is Mautic on a different server entirely to Postal?

**Comment by ffsandi** [comment]
I don't think my mautic is the problem, because before this I am using
amazon ses via smtp, and there is no problem sending big number of email up
to 100k per day.

The postal is running on separate and clean vps as your doc and I already
did some tips from this forum. Such as increase worker, threads, and edit
mysql configuration.

Thanks

**Comment by willpower232** [comment]
I assume Postal and Mautic are on different servers with the same company?
Aside from your worker processes, how much free memory does your Postal server have?
Is Postals database on the same server as Postal?

**Comment by ffsandi** [comment]
Postal & mautic are on different vps from same companies.

My free ram around 3-4 from 8gb total.

Postal database on same server.

**Comment by willpower232** [comment]
Actually I guess the most relevant question would be how any CPU threads are available.

You could get this information from neofetch or htop if you have either of those installed.

**Comment by ffsandi** [comment]
There is no problem about CPU usage, it's only about around 5-10%

I am regularly checking the usage via htop.

**Reply by willpower232** [reply]
and in htop, how many CPU bars are there at the top?

**Comment by ffsandi** [comment]
4 CPU bar

I recently reinstall postal with v2 version, just want to make sure if
there will be better performance.

I'll let you know once I did the test sending using the v2.

**Reply by adamcooke** [reply]
To scale up, there are some things you can do:

* Provide plenty of RAM and CPU cores
* Optimise your database
* Add additional workers (`postal start --scale worker=4`)

**Reply by ffsandi** [reply]
Hi, @adamcooke I already scale worker as you said. I even scale it up to 20 workers, it only gives me result around 40emails/minute.

Quite the same performance with v1 installation.

How I could find MySQL config and what do I need to do to optimise my database?

**Reply by willpower232** [reply]
You've already mentioned the database was installed on the same server, is it a docker container or is it just installed normally?

**Reply by Toni202012569** [reply]
Good afternoon. I have the same problem. The sending speed is very slow, only 200 emails per minute. How to ultimately increase your mailing speed. My RAM is at 50% capacity. Does this affect the speed of email delivery? Or will the effect only occur at 100% RAM load?

**Reply by akhil** [reply]
I have the same problem as well. RAM at 10%, CPU at 30% after scaling up. Barely doing 50 emails per minute

**Reply by sandrovenezuela** [reply]
Hi,

It may be a limitation of the data center used. For example, Hostinger allows you to create a Postal server, but you cannot use it for bulk email.

In other words, there is a limit on how many emails can be sent per minute.

**Reply by akhil** [reply]
The provider assured me there is no throttling enforced by them.

**Reply by gitbnitin** [reply]
Yes, I feel the same thing since we upgraded to 3.x.x release. the sending rate not going beyond 40, even when 80% of the resources like CPU and Memory and Disk I/O are free. Previously, it was sending more than 200/minutes. Something is wrong post the removal of the queue, but I am not sure how to demo it to the respected development team.

**Comment by apradip** [comment]
Hi guys!

I have found a solution:

Investigation: Performance Bottleneck in Postal Delivery

The Problem
Our SMTPProxy service was enqueuing messages in ~4ms, but the Postal server was experiencing a consistent 30-second delay per delivery attempt.

 Root Cause Analysis
After SSHing into the Postal node and inspecting the worker logs (docker logs postal-worker-1), we identified the following error:

```bash
ERROR Cannot connect to 2404:6800:4003:c04::1b:25 (smtp.google.com) (Net::OpenTimeout: execution expired)
```

**The issue was an IPv6 Routing Timeout:**

The Postal worker was attempting to connect to remote SMTP servers (specifically Gmail) via IPv6 because the local system had link-local IPv6 addresses enabled.
However, the server lacked a global IPv6 gateway/route, causing the connection to hang for exactly 30 seconds until the Net::OpenTimeout threshold was reached.
Only after this timeout would Postal's delivery engine fail over to IPv4 and successfully send the email.

Applied Fix
To resolve this, we disabled IPv6 at the host level to force the system and its Docker containers to prefer IPv4 immediately.

1. Applied sysctl changes:

```
bash
# Permanently disabled IPv6 in /etc/sysctl.conf
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
```

2. Applied configuration:

```bash
sysctl -p
```

3. Restarted Services: Restarted the postal-worker-1 and postal-caddy containers to ensure the networking stack picked up the changes.

✅ Verification
Log Audit: Monitored logs for 5+ minutes post-restart. The Net::OpenTimeout errors have completely ceased.
Latency Check: Outbound SMTP handshakes now occur instantly over IPv4, eliminating the 30-second execution delay.

**Reply by apradip** [reply]
If your VPS doesnt comes with default IPv6 then you will be facing this delay.

**Reply by Erwane** [reply]
Could you try with your IPv6 enabled and create an IP Pool with only your IPv4 and use this pool by default ?

**Reply by apradip** [reply]
There was an issue related to IP POOL as well.

IP Pool Rotation Failure (The "Single-IP Lock")
### Problem:
Postal only uses one IP despite having a pool. This is caused by two layers of misconfiguration:
1.  **Kernel Layer**: The Linux kernel blocks "spoofed" outgoing IPs by default.
2.  **Database Layer**: Mail servers default to `ip_pool_id = NULL`, forcing use of the system's primary IP only.

### Fix: Enable Rotation
- **Kernel Action**: Enable `net.ipv4.ip_nonlocal_bind = 1`.
- **DB Action**: Associate your mail servers with your IP Pool (Pool ID: 2).
- **Command**: Included in `scripts/postal-master-fix.sh`.
